### PR TITLE
commands: Replace all characters in selection

### PR DESCRIPTION
**Pull request by wojciechkepka**
Closes: #161

### CODE REVIEW

File: helix-term/src/commands.rs [hunk 1]
```diff
@@ -435,7 +435,11 @@ pub fn replace(cx: &mut Context) {
 
             let transaction =
                 Transaction::change_by_selection(doc.text(), doc.selection(view.id), |range| {
-                    (range.from(), range.to() + 1, Some(text.clone()))
+                    (
+                        range.from(),
+                        range.to() + 1,
```
Comment (archseer) [inline helix-term/src/commands.rs:440]: That was quick! Since this function doesn't have proper bounds checks, can you add that? Should be identical to https://github.com/helix-editor/helix/blob/0201ef920516db0fd4a4e6939f87bd0d7e4cf5dd/helix-term/src/commands.rs#L794-L796

Comment (wojciechkepka) [reply helix-term/src/commands.rs]: Done :)

**Comment by wojciechkepka** [comment]
I pushed some changes, previously `r` would place all the replaced characters on one line as no new line characters were added. Now it properly replaces all characters but newlines in selection. Is this correct behaviour or should I further adjust it?

**Comment by archseer** [comment]
Good question, kakoune seems to replace newlines too but maybe we can change the behavior here.

\cc @CBenoit @pickfire any thoughts on this?

File: helix-term/src/commands.rs [hunk 2]
```diff
@@ -429,13 +429,20 @@ pub fn replace(cx: &mut Context) {
...
+                    let max_to = doc.text().len_chars().saturating_sub(1);
+                    let to = std::cmp::min(max_to, range.to() + 1);
+                    let text: String = doc
+                        .text()
+                        .slice(range.from()..to)
+                        .chars()
+                        .map(|c| if c == '\n' { '\n' } else { ch })
+                        .collect();
+
+                    (range.from(), to, Some(text.as_str().into()))
```
Comment (pickfire) [inline helix-term/src/commands.rs:445]: Do we need to convert to `String` first and then convert to `&str` and then to `Rope`? Why not just directly convert into `Rope` or `&str`?

**Comment by pickfire** [comment]
I never had to replace across lines so I don't have hard opinion on this.

File: helix-term/src/commands.rs (hunk 2 above)
Comment (archseer) [reply helix-term/src/commands.rs:445]: The parameter is an `Option<Tendril>`, a small string abstraction (uses stack for short lengths, will heap allocate if necessary). I think the `as_str()` call isn't needed here, `into()` might be enough.

**Comment by CBenoit** [comment]
To be honest, my main use for `r`eplace is when I want to change one specific character, so I'm not completely sure. If anything, I think ignoring newlines play better with line-wise selection. For example in kakoune if I want to change all characters on one line, I select with `x` and then replace with `r{char}`. However, `x` will select the newline character that it will then get replaced and gives us an additional (and probably unwanted) `{char}`.

Replacing a newline could be useful to change it for a space effectively giving us a join operation, but we already have a `j`oin command better suited for that.

TL;DR: I'm rather in favor of ignoring the newlines.

By the way, (neo)vim also ignores newlines.

**Comment by wojciechkepka** [comment]
That is also what made me add this changes, I noticed that when I select a whole line and replace it with something the newline character also gets replaced and the next line is appended to the previously selected line.

File: helix-term/src/commands.rs (hunk 2 above)
Comment (wojciechkepka) [reply helix-term/src/commands.rs:445]: That's true, the `as_str` called can be omitted. I was pretty sure that it gave me a compilation error when I tried this before.
